testcases/open_posix_testsuite/conformance/interfaces/strchr/1-1.c

```diff
@@ -0,0 +1,103 @@
+ *	considered to be part of the string. Upon completion, the function will
+ *	return a pointer to the byte, or a null pointer if the byte was not found.
+ *
+ * method:
+ *	-Generate sample string s1
+ *	-Define one char, taking a character from the sample string(char1), one char as NULL(char2)
+	and third character out of the sample string(char3).
+ *	-Use strchr for string with char1 and store result in result1.
+ *	-Use strchr for string with char2 and store result in result2.
+ *	-Use strchr for string with char3 and store result in result3.
+ *	-Compare the result1 with pointer to the defined char(char1).
+ *	-Compare the result2 with pointer to the defined char(char2).
+ *	-Compare result3 with NULL.
+ *	-Repeat the above all steps for given number of iterations.
+
+*/
+
+#define _XOPEN_SOURCE 700
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include "posixtest.h"
+
+#define STRING_MAX_LEN 50000
+#define STEP_COUNT 2000
+#define TNAME "strchr/1-1.c"
+
+char *random_string(int len)
+{
+	int i;
+	char *output_string = malloc(len + 1);
+	if (output_string == NULL) {
+		printf(TNAME " failed to allocate memory\n");
+		exit(PTS_UNRESOLVED);
+	}
+	for (i = 0; i < len; i++) {
+		output_string[i] = rand() % 254 + 1;
+			//Avoid adding 'n' and '\0' character in the string
+			if (output_string[i] == 'n' || output_string[i] == '\0')
+				i--;
+	}
+	output_string[len] = '\0';
+	return output_string;
+}
+
+int main(void)
+{
+	int i, char_pos;
+
+	for (i = 10; i < STRING_MAX_LEN; i += STEP_COUNT) {
+		char *sample_str;
+		char sample_char_1;
+		char sample_char_2 = '\0';
+		char sample_char_3 = 'n';
+		char_pos = rand() % 10;
+
+		sample_str = random_string(i);
+		sample_char_1  = sample_str[char_pos];
+
+		char *ret_str_1 = strchr(sample_str, sample_char_1);
+
+		char *ret_str_2 = strchr(sample_str, sample_char_2);
+
+		char *ret_str_3 = strchr(sample_str, sample_char_3);
+
+		if(ret_str_1 != &sample_str[char_pos]) {
+			printf(TNAME " Test Failed, Failed to return pointer to the byte, when matching"
+					" character is found. Expected: %p, but returned: %p\n",
+					&sample_str[char_pos], ret_str_1);
+			exit(PTS_FAIL);
+		} else if (ret_str_2 != &sample_str[i]) {
+			printf(TNAME " Test Failed, Failed to consider Null character as a part of the"
+					" string,\n\t\t\t\t   so failed to return pointer to the NULL byte,"
+					" when matching character is found.\n\t\t\t\t   Expected: %p,"
+					" but returned: %p\n", &sample_str[i - 1], ret_str_2);
+		} else if (ret_str_3 != NULL) {
+			printf(TNAME " Test Failed, Failed to return NULL when character is not found"
+					" in the string. Expected: NULL, but returned: %p\n", ret_str_3);
+			exit(PTS_FAIL);
+		}
+		free(sample_str);
+	}
+	printf(TNAME " Test Passed, First case: string with char1, byte is found and"
+				" so returned pointer to the byte.\n\t\t\t  Second case: The"
+				" terminating NULL character is considered as part of the string,"
```

Comment:
This section's confusing. `NULL` (all caps) typically refers to a null pointer, which has a specific definition per POSIX (`(void*)0`). With this in mind, `NULL byte` is confusing/incorrect, because `sizeof(void*) == sizeof(long long)`, which is 4 bytes on 32-bit platforms and 8 bytes on 64-bit platforms.

I typically use `null character` to refer to `\0`, but opengroup.org uses the term `null byte` in the definition, so I think sticking with their terminology is the way to go.